src/ffigen/ffigen.py

```diff
@@ -172,6 +173,54 @@ def need_cwrap(id):
         )
         print(js_declaration)
 
+    # Generate functions to read/write simple types.
+    for type in simple_type_mapping.values():
+        if type == "void":
+            continue
+        print(f"const _sh_ptr_read_{type} = $SHBuiltin.extern_c({{}}, function _sh_ptr_read_{type}(ptr: c_ptr, offset: c_int): {type} {{ throw 0; }});")
+        print(f"const _sh_ptr_write_{type} = $SHBuiltin.extern_c({{}}, function _sh_ptr_write_{type}(ptr: c_ptr, offset: c_int, v: {type}): void {{}});")
+
+    print(f"const _sh_ptr_add = $SHBuiltin.extern_c({{}}, function _sh_ptr_add(ptr: c_ptr, offset: c_int): c_ptr {{ throw 0; }});")
+
+    # Generate struct and union sizes.
+    for type in ["Struct", "Union"]:
+        for elem in root.findall(".//" + type):
+            name = elem.get("name")
+            size = int(elem.get("size")) // 8
+            print(f"const _sizeof_{name} = {size};")
+
+    # Generate struct accessors
+    for field in root.findall(".//Field"):
+        field_name = field.get("name")
+        field_type_id = field.get("type")
+        field_type_elem = id_to_element[field_type_id]
+
+        struct_name = id_to_element[field.get("context")].get("name")
+        sh_type = to_sh_name(field_type_id)
+
+        offset = int(field.get("offset")) // 8
+
+        # If the field is a struct, union, or array, return a pointer to 
+        # it and do not emit a setter.
+        if field_type_elem.tag == "Array" or field_type_elem.tag == "Union" or field_type_elem.tag == "Struct":
```

Comment:
Need a constant constant containing the size of the struct.

Reply:
They are emitted above on line 186.